src/program/frame.rs

```diff
@@ -35,18 +35,57 @@ impl FrameSet {
         self.frames.keys().collect()
     }
 
-    /// Return all contained FrameIdentifiers which include **exactly** these qubits (in any order)
-    /// and - if names are provided - match one of the names.
-    pub fn get_matching_keys(&self, qubits: &[Qubit], names: &[String]) -> Vec<&FrameIdentifier> {
-        let qubit_set: HashSet<&Qubit> = qubits.iter().collect();
-        self.frames
-            .iter()
-            .filter(|(identifier, _)| {
-                (names.is_empty() || names.contains(&identifier.name))
-                    && qubit_set == identifier.qubits.iter().collect()
-            })
-            .map(|(id, _)| id)
-            .collect::<Vec<_>>()
+    /// Return all frames in the set which match all of these conditions. If a frame _would_ match, but is
+    /// not present in this [FrameSet], then it is not returned (notably, the [FrameMatchCondition::Specific]
+    /// match condition).
+    pub(crate) fn get_matching_keys<'s, 'a>(
+        &'s self,
+        condition: FrameMatchCondition<'a>,
+    ) -> HashSet<&'s FrameIdentifier> {
+        let keys = self.frames.keys();
+
+        match condition {
+            FrameMatchCondition::All => keys.collect(),
+            FrameMatchCondition::AnyOfNames(names) => {
+                keys.filter(|&f| names.contains(&f.name)).collect()
+            }
+            FrameMatchCondition::AnyOfQubits(qubits) => {
+                let any_of_set: HashSet<_> = qubits.iter().collect();
+
+                keys.filter(|&f| f.qubits.iter().any(|q| any_of_set.contains(q)))
+                    .collect()
+            }
+            FrameMatchCondition::ExactQubits(qubits) => {
+                let all_of_set: HashSet<_> = qubits.iter().collect();
+
+                keys.filter(|&f| f.qubits.iter().all(|q| all_of_set.contains(q)))
+                    .collect()
+            }
+            FrameMatchCondition::Specific(frame) => {
+                // This unusual pattern (fetch key & value by key, discard value) allows us to return
+                // a reference to `self` rather than `condition`, keeping lifetimes simpler.
+                if let Some((frame, _)) = self.frames.get_key_value(frame) {
+                    vec![frame].into_iter().collect()
+                } else {
+                    HashSet::new()
+                }
+            }
+            FrameMatchCondition::And(conditions) => conditions
+                .into_iter()
+                .map(|c| self.get_matching_keys(c))
+                .reduce(|acc, el| acc.into_iter().filter(|&v| el.contains(v)).collect())
```

Comment:
I suppose my go-to reaction shouldn't be "maybe we should add another dependency!" but I think `any` and `all` from [`itertools`](https://docs.rs/itertools/latest/itertools/) might be useful.

Reply:
Yeah I think for only a single use case here I might prefer to keep out the added dep

Reply:
For sure.